### ISSUE TITLE
Add support to generate plugin extensions

### ISSUE THREAD

**Issue by JavierSegoviaCordoba** [GradleUp/gratatouille#1]
The extensions are isolated so they can be seen as a data class to contain all properties and any behavior on it can be a method.

Any knowledge about lazy APIs and Gradle internal could be ignored and just use Kotlin basic types, under the hood they will be mapped to lazy APIs.

An example could be:

```kotlin
@GExtension
data class MyFancyExtension(
    val prop1: String,
    val prop2: GFile, // or `GInputFile` ?
    @GAssisted val randomThing: RandomThing,
) {
    fun foo() {
         // do something with props
    }
}
```

```kotlin
// generated, automatically add `ProjectLayout`, `ObjectFactory`, 
// and any `@GAssisted` value.
open class MyFancyExtension @Inject constructor(...) {

    val prop1: Property<String> = ...

    val prop2: FileProperty = ...

    fun foo() {
        // do something with props
    }
}
```

The annotation `@GAssisted` would help to enable doing anything in complex use cases.

Nested extensions should be supported automatically:

```kotlin
@GExtension
data class RootFancyExtension(val nestedExtension: NestedExtension, ...)

@GExtension
data class NestedFancyExtension(...)
```

should be consumed on the user side:

```kotlin
rootFancy {
    // ...
    
    nestedFancy {
        // ...
    }
}
```

**Comment by martinbonnin** [comment]
@JavierSegoviaCordoba latest versions have a [`@GExtension`](https://github.com/GradleUp/gratatouille/blob/main/gratatouille-runtime/src/main/kotlin/gratatouille/api.kt#L91) annotation, I'm not sure it's 100% what you wanted though.

You can see it in action [here](https://github.com/apollographql/apollo-kotlin-faker/blob/e6fc8068bee5dcff79fbe71ff3a98e8f00f0567a/gradle-plugin/src/main/kotlin/com/apollographql/faker/ApolloFakerExtension.kt#L7):

```kotlin
@GExtension(pluginId = "com.apollographql.faker")
abstract class ApolloFakerExtension(private val project: Project) {
  private val services = mutableSetOf<String>()
  fun service(name: String, action: Action<ApolloFakerService>) {
    check(!services.contains(name)) {
      "Apollo Faker: service '$name' is already created."
    }

    services.add(name)
    ApolloFakerService(project, name)
      .also(action::execute)
      .registerTasks()
  }
}
```

Basically, `@GExtension` automatically creates a plugin under the hood that calls `Project.extensions.create()` so service injection is available there. It also takes care of creating a plugin descriptor automagically. 

Let me know if that'd work for you!
